sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -775,6 +775,24 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
     }
   }
 
+  /**
+   * Spark will append a ColumnarToRow (C2R) to a GpuDataWritingCommandExec, but this
+   * C2R is useless because DataWritingCommandExec returns empty data.
+   * Besides, Spark will trigger a new job for the operators after a DataWritingCommandExec,
+   * and makes the DataWritingCommandExec be the data source node for this new job. Since
+   * no data is generated, this new job is entirely a noop.
+   *
+   * This rule removes the tailing C2R to eliminate the useless job. Then
+   * GpuDataWritingCommandExec becomes the root node, and should support row-based
+   * executions to interact with Spark.
+   */
+  private def removeTailingColumnarToRow(plan: SparkPlan): SparkPlan = plan match {
+    // No need to do it recursively, because the data write command can not be a
+    // middle node of a Spark job.
+    case GpuColumnarToRowExec(child @ GpuDataWritingCommandExec(_, _), _) => child
+    case _ => plan
+  }
```

Comment:
Rather than hacking the plan to remove a columnar to row, why don't we avoid putting it in the plan in the first place?  Similar to GpuColumnarToRowExec, we could have GpuDataWritingCommandExec say it does _not_ support columnar execution (even though it does) and have it inherit ColumnarToRowTransition (because that's sorta what it's doing here).  This approach is described in #932.

Reply:
Good to know.
I tried this suggestion, but it leads to some errors for some cases at least on Spark 321.
test query:
```
spark.read.parquet("/data/tmp/data-hive-test/t1").filter("ca != null").write.mode("overwrite").parquet("/data/tmp/out/")
```

errors:
```
Caused by: java.lang.IllegalStateException: Internal Error class org.apache.spark.sql.execution.LocalTableScanExec has column support mismatch:
LocalTableScan <empty>, [ca#36, cb#37L, cc#38, cd#39]

  at org.apache.spark.sql.execution.SparkPlan.doExecuteColumnar(SparkPlan.scala:312)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:211)
  at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:222)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
```

According to [the rule of Spark 321 here](https://github.com/apache/spark/blob/v3.2.1/sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala#L528), it will not be applied to the children if the root node does not support columnar and is a ColumnarToRowTransition.

Anyway, this is a short-term fix.

Reply:
Why are we doing a short-term fix?  This is a long-standing issue, wondering why the sudden need to hack this?  This is leaving the plan in a state where Spark really doesn't want that transition removed.  It put it there because we're not inheriting from ColumnarToRowTransition and thus Spark thinks that transition needs to be there.